src/test/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilderTest.java

```diff
@@ -0,0 +1,147 @@
+import static org.testng.Assert.assertNull;
+
+import static com.github.tomakehurst.wiremock.client.WireMock.equalTo;
+import static com.github.tomakehurst.wiremock.client.WireMock.postRequestedFor;
+
+
+public class ClickHouseHttpClientBuilderTest {
+
+    private static WireMockServer mockServer;
+
+    @BeforeClass
+    public static void beforeAll() {
+        mockServer = new WireMockServer(
+            WireMockConfiguration.wireMockConfig().dynamicPort());
+        mockServer.start();
+    }
+
+    @AfterMethod
+    public void afterTest() {
+        mockServer.resetAll();
+    }
+
+    @AfterClass
+    public static void afterAll() {
+        mockServer.stop();
+        mockServer = null;
+    }
+
+    @Test
+    public void testCreateClientContextNull() {
+        assertNull(ClickHouseHttpClientBuilder.createClientContext(null).getAuthCache());
+    }
+
+    @Test
+    public void testCreateClientContextNoUserNoPass() {
+        assertNull(ClickHouseHttpClientBuilder.createClientContext(new ClickHouseProperties())
+            .getAuthCache());
+    }
+
+    @Test
+    public void testCreateClientContextNoHost() {
+        ClickHouseProperties props = new ClickHouseProperties();
+        props.setUser("myUser");
+        props.setPassword("mySecret");
+        assertNull(ClickHouseHttpClientBuilder.createClientContext(props).getAuthCache());
+    }
+
+    @Test
+    public void testCreateClientContextUserPass() {
+        ClickHouseProperties props = new ClickHouseProperties();
+        props.setUser("myUser");
+        props.setPassword("mySecret");
+        props.setHost("127.0.0.1");
+        assertEquals(
+            ClickHouseHttpClientBuilder.createClientContext(props).getAuthCache()
+                .get(HttpHost.create("http://127.0.0.1:80")).getSchemeName(),
+            "basic");
+    }
+
+    @Test
+    public void testCreateClientContextOnlyUser() {
+        ClickHouseProperties props = new ClickHouseProperties();
+        props.setUser("myUser");
+        props.setHost("127.0.0.1");
+        assertEquals(
+            ClickHouseHttpClientBuilder.createClientContext(props).getAuthCache()
+                .get(HttpHost.create("http://127.0.0.1:80")).getSchemeName(),
+            "basic");
+    }
+
+    @Test
+    public void testCreateClientContextOnlyPass() {
+        ClickHouseProperties props = new ClickHouseProperties();
+        props.setPassword("myPass");
+        props.setHost("127.0.0.1");
+        assertEquals(
+            ClickHouseHttpClientBuilder.createClientContext(props).getAuthCache()
+                .get(HttpHost.create("http://127.0.0.1:80")).getSchemeName(),
+            "basic");
+    }
+
+
+    @Test(dataProvider = "authUserPassword")
+    public void testHttpAuthParametersCombination(String authorization, String user,
+        String password, String expectedAuthHeader) throws Exception
+    {
+        ClickHouseProperties props = new ClickHouseProperties();
+        props.setHost("localhost");
```

Comment:
Could you add one more test when host is null?

Reply:
Sure

Reply:
Thought host is null when using BalancedDataSource but I'll double check :p